### PR TITLE
Reposupport

### PR DESCRIPTION
**Pull request by Atul-source**
This PR is for issue #18

Before you could only have a single repository defined for all eBPF programs in the global config.  This change enables the ability to define an eBPF repository per program.

### CODE REVIEW

File: models/l3afd.go [hunk 1]
```diff
@@ -23,32 +23,33 @@ type L3afDNFArgs map[string]interface{}
...
+	Rules             string              `json:"rules"`                  // Config rules
+	ConfigFilePath    string              `json:"config_file_path"`       // Config file location
+	CfgVersion        int                 `json:"cfg_version"`            // Config version
+	StartArgs         L3afDNFArgs         `json:"start_args"`             // Map of arguments to start command
+	StopArgs          L3afDNFArgs         `json:"stop_args"`              // Map of arguments to stop command
+	StatusArgs        L3afDNFArgs         `json:"status_args"`            // Map of arguments to status command
+	MapArgs           L3afDNFArgs         `json:"map_args"`               // Config BPF Map of arguments
+	ConfigArgs        L3afDNFArgs         `json:"config_args"`            // Map of arguments to config command
+	MonitorMaps       []L3afDNFMetricsMap `json:"monitor_maps"`           // Metrics BPF maps
+	DownloadUrl       string              `json:"ebpf_program_repo_name"` // Download url for Program
```
Comment (satyaranjanp) [inline models/l3afd.go:52]: Can we rename the variable to something more related to eBPF Package repository, for ex. EPRUrl. Also, can we have the json tag as ebpf_package_repo_url

Comment (satyaranjanp) [inline models/l3afd.go:52]: Can we rename the variable to something more related to eBPF Package repository, for ex. EPRUrl. Also, can we have the json tag as ebpf_package_repo_url

Comment (Atul-source) [reply models/l3afd.go]: I have changed variables names

File: kf/bpf.go [hunk 2]
```diff
@@ -523,40 +523,79 @@ func (b *BPF) VerifyAndGetArtifacts(conf *config.Config) error {
 // GetArtifacts downloads artifacts from the nexus repo
 func (b *BPF) GetArtifacts(conf *config.Config) error {
 	var fPath = ""
+	buf := &bytes.Buffer{}
+	//      "file://" ---> has size 7
+	//     "Http://" ---> has size 7
```
Comment (dthaler) [inline kf/bpf.go:528]:
```suggestion
	//     "file://" ---> has size 7
	//     "http://" ---> has size 7
```

**Review by dthaler — COMMENTED** [review]
Thanks for doing this!   Changes look good so far.

https://github.com/l3af-project/l3afd/blob/main/docs/CONTRIBUTING.md#guidelines-for-pull-requests says:

> Describe the tests that you ran to verify your changes and add/update test cases
> Update relevant docs, especially if you've changed APIs

and also

> PRs are expected to have 100% test coverage for added code. This can be verified with a coverage build. If your PR cannot have 100% coverage for some reason please clearly explain why when you open it. 

What about documentation changes for this?
What about tests that verify that local files work?  (PR #66 added some tests, but since this PR is separate I assume that PR doesn't cover tests for this PR)

**Comment by sanfern** [comment]
@Atul-source Can you update api documentation https://github.com/l3af-project/l3afd/blob/main/docs/api/README.md

File: docs/api/README.md [hunk 3]
```diff
@@ -15,6 +15,7 @@ The payload will look more like this standard JSON:
           "name": "ratelimiting",
           "seq_id": 1,
           "artifact": "l3af_ratelimiting.tar.gz",
+          "ebpf_package_repo_url": "http://www.exampleebpf.com/l3af_ratelimiting.tar.gz"
```
Comment (jniesz) [inline docs/api/README.md:18]: Let's put example URL of domain we own, such as l3af.io

File: kf/bpf.go [hunk 4]
```diff
@@ -523,40 +523,79 @@ func (b *BPF) VerifyAndGetArtifacts(conf *config.Config) error {
 // GetArtifacts downloads artifacts from the nexus repo
 func (b *BPF) GetArtifacts(conf *config.Config) error {
 	var fPath = ""
+	buf := &bytes.Buffer{}
+	//      "file://" ---> has size 7
+	//     "http://" ---> has size 7
+	if len(b.Program.EPRUrl) >= 7 {
```
Comment (jniesz) [inline kf/bpf.go:529]: We should just use url.parse to determine the scheme (http, https, file, ...).  File should have three slashes "file:///"

File: kf/bpf.go [hunk 5]
```diff
@@ -522,137 +525,182 @@ func (b *BPF) VerifyAndGetArtifacts(conf *config.Config) error {
...
+				if err == io.EOF {
+					break
+				} else if err != nil {
+					return fmt.Errorf("untar failed: %w", err)
+				}
 
-			if strings.Contains(header.Name, "..") {
-				return fmt.Errorf("zipped file contians filepath (%s) that includes (..)", header.Name)
-			}
+				if strings.Contains(header.Name, "..") {
```
Comment (jniesz) [inline kf/bpf.go:672]: Can we have a similar check as the zip filename check above with filepath.Clean() ?  It probably also makes breaking out this code, so as we add support for new compression algorithms the filename checking logic remains the same and can be managed in single place.

**Review by jniesz — COMMENTED** [review]
Left some comments for changes

File: kf/bpf.go [hunk 6]
```diff
@@ -522,137 +525,182 @@ func (b *BPF) VerifyAndGetArtifacts(conf *config.Config) error {
...
+					}
+					buf.ReadFrom(f)
+					f.Close()
+				} else {
+					return fmt.Errorf("artifact is not found")
+				}
+			}
+		}
+	} else {
+		kfRepoURL, err := url.Parse(conf.KFRepoURL)
```
[the remark this answers is not in the document]
Comment (Atul-source) [reply kf/bpf.go]: I will do a separate PR for a change of KFRepoURL --> eBPFRepoURL

File: kf/bpf.go (hunk 5 above)
Comment (Atul-source) [reply kf/bpf.go]: if we check for somewhere else using the helper function it is giving security issues.

Comment (jniesz) [reply kf/bpf.go]: breaking out into a separate function vs. having similar code 2+ more times seems to make sense, not sure why this is alerting if we make this change.  We can discuss more on call

File: docs/api/README.md [hunk 7]
```diff
@@ -46,7 +47,10 @@ The payload will look more like this standard JSON:
...
 | seq_id              | number                                         | `1`                                                            | Position of the eBPF program in the chain. Count starts at 1.                                                                    |
-| artifact            | string                                         | `"l3af_ratelimiting.tar.gz"`                                   | Userspace eBPF program binary and kernel eBPF byte code in tar.gz format                                                         |
+| artifact            | string                                         | `"l3af_ratelimiting.tar.gz"`                                   | Userspace eBPF program binary and kernel eBPF byte code in tar.gz format     
+|
+| ebpf_package_repo_url | string         | `"https://l3af.io/"`     | eBPF package repository URL.  If it is not provided default URL is used.
+|                                                  |
```
Comment (dthaler) [inline docs/api/README.md:53]:
```suggestion
| artifact            | string                                         | `"l3af_ratelimiting.tar.gz"`                                   | Userspace eBPF program binary and kernel eBPF byte code in tar.gz format     |
| ebpf_package_repo_url | string         | `"https://l3af.io/"`     | eBPF package repository URL.  If it is not provided default URL is used. |                                                  |
```

Comment (dthaler) [reply docs/api/README.md:53]: The preview looks malformed, maybe due to the extra newlines (not sure)

**Review by jniesz — APPROVED** [review]
LGTM

File: kf/bpf_test.go [hunk 8]
```diff
@@ -325,12 +336,28 @@ func TestBPF_isRunning(t *testing.T) {
...
 
+// RoundTripFunc .
+type RoundTripFunc func(req *http.Request) *http.Response
+
+// RoundTrip .
+func (f RoundTripFunc) RoundTrip(req *http.Request) (*http.Response, error) {
+	return f(req), nil
+}
+
+//NewTestClient returns *http.Client with Transport replaced to avoid making real calls
```
Comment (dthaler) [inline kf/bpf_test.go:347]:
```suggestion
// NewTestClient returns *http.Client with Transport replaced to avoid making real calls
```

**Review by dthaler — APPROVED** [review]

**Review by sanfern — APPROVED** [review]

**Pull request merged**